File: database/src/main/java/org/difft/app/database/models/ForwardModel.java
package org.difft.app.database.models;

import com.tencent.wcdb.WCDBField;
import com.tencent.wcdb.WCDBIndex;
import com.tencent.wcdb.WCDBTableCoding;

import java.util.Objects;

@WCDBTableCoding
public class ForwardModel {
    @WCDBField(isPrimary = true, isAutoIncrement = true)
    public long databaseId;

    @WCDBIndex
    @WCDBField
    public long id;

    @WCDBField
    public int type;

    @WCDBField
    public boolean isFromGroup = false;

    @WCDBField
    public String author = "";

    @WCDBField
    public String text = "";

    @WCDBField
    public long serverTimestamp = 0L;

    @WCDBField
    public Long cardModelDatabaseId;

    @WCDBField
    public Long parentForwardModelDatabaseId;

    @WCDBField
    public Long forwardContextDatabaseId;

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ForwardModel that = (ForwardModel) o;
        return id == that.id && type == that.type && isFromGroup == that.isFromGroup && serverTimestamp == that.serverTimestamp && Objects.equals(author, that.author) && Objects.equals(text, that.text) && Objects.equals(cardModelDatabaseId, that.cardModelDatabaseId) && Objects.equals(parentForwardModelDatabaseId, that.parentForwardModelDatabaseId) && Objects.equals(forwardContextDatabaseId, that.forwardContextDatabaseId);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, type, isFromGroup, author, text, serverTimestamp, cardModelDatabaseId, parentForwardModelDatabaseId, forwardContextDatabaseId);
    }
}

File: database/src/main/java/org/difft/app/database/models/ForwardModel.kt
package org.difft.app.database.models

import com.tencent.wcdb.WCDBField
import com.tencent.wcdb.WCDBIndex
import com.tencent.wcdb.WCDBTableCoding
import java.util.Objects

@WCDBTableCoding
class ForwardModel {
    // Int (not Long): WCDB-KSP generates `== 0` autoincrement guard which does not
    // compile for Long primary keys. Column type stays ColumnType.Integer — no schema
    // change. Local rowid never exceeds 2^31 and is never externally serialized. #901
    @WCDBField(isPrimary = true, isAutoIncrement = true)
    var databaseId: Int = 0

    @WCDBIndex
    @WCDBField
    var id: Long = 0L

    @WCDBField
    var type: Int = 0

    @WCDBField
    var isFromGroup: Boolean = false

    @WCDBField
    var author: String = ""

    @WCDBField
    var text: String = ""

    @WCDBField
    var serverTimestamp: Long = 0L

    @WCDBField
    var parentForwardModelDatabaseId: Long? = null

    @WCDBField
    var forwardContextDatabaseId: Long? = null

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        other as ForwardModel
        return id == other.id &&
                type == other.type &&
                isFromGroup == other.isFromGroup &&
                serverTimestamp == other.serverTimestamp &&
                author == other.author &&
                text == other.text &&
                parentForwardModelDatabaseId == other.parentForwardModelDatabaseId &&
                forwardContextDatabaseId == other.forwardContextDatabaseId
    }

    override fun hashCode(): Int =
        Objects.hash(
            id,
            type,
            isFromGroup,
            author,
            text,
            serverTimestamp,
            parentForwardModelDatabaseId,
            forwardContextDatabaseId
        )
}
